utils/gnutls_offset.sh

```diff
@@ -0,0 +1,66 @@
+#!/usr/bin/env bash
+set -e
+
+PROJECT_ROOT_DIR=$(pwd)
+GNUTLS_DIR="${PROJECT_ROOT_DIR}/deps/gnutls"
+OUTPUT_DIR="${PROJECT_ROOT_DIR}/kern"
+
+if [[ ! -f "go.mod" ]]; then
+  echo "Run the script from the project root directory"
+  exit 1
+fi
+
+echo "check file exists: ${GNUTLS_DIR}/.git"
+# skip cloning if the header file of the max supported version is already generated
+if [[ ! -f "${GNUTLS_DIR}/.git" ]]; then
+  echo "check directory exists: ${GNUTLS_DIR}"
+  # skip cloning if the gnutls directory already exists
+  if [[ ! -d "${GNUTLS_DIR}" ]]; then
+    echo "git clone gnutls to ${GNUTLS_DIR}"
+    git clone https://github.com/gnutls/gnutls.git ${GNUTLS_DIR}
+  fi
+fi
+
+
+function run() {
+  git fetch --tags
+  cp -f ${PROJECT_ROOT_DIR}/utils/gnutls_offset.c ${GNUTLS_DIR}/offset.c
+  main_version="3.8"
+
+  for ver in $(seq 7 8); do
+    tag="${main_version}.${ver}"
+    underline_tag=$(echo $tag | tr "." "_")
+    header_file="${OUTPUT_DIR}/gnutls_${underline_tag}_kern.c"
+    header_define="GNUTLS_${underline_tag}_KERN_H"
+
+    if [[ -f ${header_file} ]]; then
+      echo "Skip ${header_file}"
+      continue
+    fi
+    echo "git checkout ${tag}"
+    git checkout ${tag}
+    echo "Generating ${header_file}"
+
+    # init
```

Comment:
During initialization, some class libraries are missing and need to be installed in advance.

```shell
sudo apt install gnulib libtool gtk-doc-tools nettle-dev libtasn1-6-dev libunistring-dev libev-dev libunbound-dev
```

Reply:
Around line 43, some warnings are output? If an error is reported during execution, please ensure that the class library is installed successfully, or refer to [gnutls github](https://github.com/gnutls/gnutls)

Reply:
`gnulib` is no needed, because the gnutls will use it as a submodule.

Reply:
```shell
Submodule path 'devel/libtasn1': checked out '2b7ee166f031bea63c5e7badcedc69f1cea69c89'
./bootstrap: 69: ./bootstrap.conf: gnulib/gnulib-tool: not found
./bootstrap: Bootstrapping from checked-out gnutls sources...
./bootstrap: consider installing git-merge-changelog from gnulib
./bootstrap: getting gnulib files...
Submodule 'gnulib' (https://gitlab.com/libidn/gnulib-mirror.git) registered for path 'gnulib'
Cloning into '/home/ubuntu/ecapture/deps/gnutls/gnulib'...
```

I tested it on an amd64 Ubuntu 22.04 system, and it seems that I still need to clone this repository.